### PR TITLE
Admin route

### PR DESCRIPTION
**Pull request by sarahelsaig**
## Problem

The stock ASP.NET Core `[Route(template)]` attribute can only be used for front-end routes, because you can't pass the admin prefix to it. So you are forced to configure the route in the module's `Startup` like this this:
```csharp
                routes.MapAreaControllerRoute(
                    name: descriptor.DisplayName,
                    areaName: descriptor.RouteValues["area"],
                    pattern: _defaultAreaPattern.Replace("{action}", descriptor.ActionName),
                    defaults: new { controller = descriptor.ControllerName, action = descriptor.ActionName }
                );
```

This is cumbersome, less readable, and there is a degree of separation from the action that makes maintenance and debugging less intuitive.

## Solution

We already have the `[Admin]` attribute and the `AdminAreaControllerRouteMapper` that automatically assigns a default route with the admin prefix prepended. So why not give an optional Template property to the `AdminAttribute` to bring it in parity with the `RouteAttribute`? With this contribution you can do
```
[Admin("My/Path/{id}")]
```
which is the same as `[Route("Admin/My/Path/{id}")]` except the admin prefix is not hard coded.

Note: So far I've changed the admin routes in OrchardCore.Users as a demo. If you like this contrib, I will go through every `Configure` method and replace the routes that look like `pattern: _adminOptions.AdminUrlPrefix + "some/constant/value"` with the matching [Admin("some/constant/value")]. But there are a lot of them and I don't think this can be automated, so I wouldn't want to do this prematurely.

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:27]:
```suggestion
            if (descriptor.ControllerName != "Admin" && controllerAttribute == null && actionAttribute == null)
            {
                return false;
            }
```

**Review by Piedone — CHANGES_REQUESTED** [review]
I very much welcome this but please demo on the Tuesday meeting to get some quick feedback too.

File: src/OrchardCore/OrchardCore.Admin.Abstractions/AdminAttribute.cs
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Admin.Abstractions/AdminAttribute.cs:17]: typo

File: src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:34]: Is it behaving the same way as [Route] applied on a controller (I believe as a prefix to all actions attributes).

**Comment by MikeAlhayek** [comment]
@sarahelsaig all in favor of this.

File: src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs
Comment (sarahelsaig) [reply src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:34]: Yes, it's the same.

At first I wasn't sure how RouteAttribute works with controllers, so I've tested it:
```csharp
[Route("Sample")]
public class SampleController : Controller
{
    public IActionResult Index() => Content(nameof(Index));
    public IActionResult Other() => Content(nameof(Other));
}
```
If I try to navigate to `~/Sample/` I get this error because the same route gets applied to both actions:
image
And if I try to go to `~/Sample/Index` I get 404 error.

So it looks like the `[Route]` template is applied as-is, it doesn't become a prefix.

If I use `[Route("Sample/{action}")]` then `~/Sample/Index` and `~/Sample/Other` works. You can do the same with `[Admin]` too. I have included an advice to use the `{action}` in the template [here](https://github.com/OrchardCMS/OrchardCore/pull/15251/files#diff-0c3a28c7723c63ada3464e5065d04cdc7239c40b9539a5de5632b11ea87474fdR19).

Comment (sarahelsaig) [reply src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:34]: Turns out this is great for `AdminController`s, because you can simply add `[Admin("Users/{action}/{id?}")]` to the controller like [here](https://github.com/OrchardCMS/OrchardCore/pull/15251/files#diff-254176877b0597fca38954810afd6597eecc56718835a87df042721cc200c968R32).

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:34]: yes `[Admin("Users/{action}/{id?}")]` is great. Just be sure that it works with tenants that use prefix please. I think it';; still work. But just be sure it does. This should simplify lots of the boilerplate code we use in the startup file.

File: src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:43]: @sarahelsaig 
Usually if the action is Index, we don't append Index to the route, maybe we can add an in condition to allow us to not have to type /index in the URl for index actions.

File: src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs
Comment (sarahelsaig) [reply src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:34]: Thankfully this works:

image

image

File: src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs
Comment (sarahelsaig) [reply src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:43]: Because of all the possible optional template parts at the end (like `/{id?}`), I think it would be pretty complicated to make it reliable without the danger of accidentally creating conflicts. Also I think `[Route]` doesn't offer this functionality either, so it may be confusing for users.

I have added an action-level route to every Index where a matching route was defined in the startup. For example [this](https://github.com/OrchardCMS/OrchardCore/pull/15251/files#diff-6d5d8927f63774f4760387afbf2b1f8cd536fd8bb5566321f04aaef9b60f7dc5L54-L59) became [this](https://github.com/OrchardCMS/OrchardCore/pull/15251/files#diff-41124286db00cbe90fba652e472f4fa52b39ad3e8fbcb3ffccd77bda7ad18939R93).

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:43]: Okay. I guess we can add the admin attribute on the index action in addition to the controller. The attribute on the action will have a height order I am guess so it would take presence.

Another thing we can support if you up for it is a name property to identify the name of the route. Sometimes it's is helpful to be able to create url using a route name instead of action/controller. Also, I am not sure what descriptor.DisplayName returns but maybe we can use Name property and fallback on descriptor.DisplayName

Comment (sarahelsaig) [reply src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:46]: Here is an example of the default `descriptor.DisplayName`:
```
OrchardCore.Settings.Controllers.AdminController.Index (OrchardCore.Settings)
```
Looks like a cleaned-up version of the action's method header. In my opinion this is easier to understand when debugging than the names used in the `routes.MapAreaControllerRoute()` calls I've removed like AdminMenuList, BackgroundTasks, SearchUsers, etc. Is there any name that should be specifically preserved? I've added `AdminAttribute.RouteName` so it's possible.

Also I've prefixed the default name with "Admin Route for " to avoid confusion with the routes created by `DefaultAreaControllerRouteMapper`. Apparently you can have anything in the route name including spaces.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:46]: `descriptor.DisplayName` is good now since you added the RouteName option for those he will use it. thank you!

**Comment by MikeAlhayek** [comment]
@sarahelsaig can you please fix the merge conflict?

Also, is this PR ready to be reviewed?

**Comment by sarahelsaig** [comment]
> sarahelsaig can you please fix the merge conflict?

Ok, it's fixed. 

> Also, is this PR ready to be reviewed?

Yes, it can be reviewed. Thank you.

File: src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:47]: all you should need is this

```
name ?? descriptor.DisplayName
```

File: src/OrchardCore.Modules/OrchardCore.AdminDashboard/Controllers/DashboardController.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.AdminDashboard/Controllers/DashboardController.cs:88]: I am wondering if we should also specify the old route name in case some are using it in their app for routing. Not just here, but maybe everywhere

For example,
```
[Admin("dashboard/manage", "AdminDashboard")]
```

File: src/OrchardCore.Modules/OrchardCore.Recipes/Controllers/AdminController.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Recipes/Controllers/AdminController.cs:60]: [Admin("Recipes", "Recipes")]

File: src/OrchardCore.Modules/OrchardCore.Settings/Controllers/AdminController.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Settings/Controllers/AdminController.cs:44]: This route is redirected to a alot, its worth giving it the old `AdminSettings` name.

**Review by MikeAlhayek — COMMENTED** [review]
@sarahelsaig 
Currently, we are not migrating the existing route name in the attribute. This may not be a problem, but could become a braking change for someone. The risk is minimal, but it is there. It maybe a bit painful to go back and take the name of each route and add it to the Attribute so this way every name will have a name and we can continue  to to be able to route using route name.

File: src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs
Comment (sarahelsaig) [reply src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:47]: As I wrote above, I thought it's good to make it distinct from maps made by `DefaultAreaControllerRouteMapper`. But I'm not married to the idea so I can revert it.

File: src/OrchardCore.Modules/OrchardCore.AdminDashboard/Controllers/DashboardController.cs
Comment (sarahelsaig) [reply src/OrchardCore.Modules/OrchardCore.AdminDashboard/Controllers/DashboardController.cs:88]: This will be a lot of (non-trivial) work, especially after the fact. So before I get into it, can you think of an example how that name may be used? I can't think of one.

File: src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:47]: I think to avoid a possible breaking change, it is best to keep naming as is. The route in the attribute will clean things up, but I don't think we should avoid breaking change if we can help it.

File: src/OrchardCore.Modules/OrchardCore.AdminDashboard/Controllers/DashboardController.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.AdminDashboard/Controllers/DashboardController.cs:88]: yes. look at my next comment. 
When you specify this

```
routes.MapAreaControllerRoute(
    name: "AUniqueNameToTheRoute", // This is the route name. This value is something we can use to route the user to
    // ....
);
```

You can then use that value to redirect. For example `RedirectToRoute("AUniqueNameToTheRoute")`. If someone in their project is already doing route name routing, their code will break after merging this PR.

**Comment by sarahelsaig** [comment]
@MikeAlhayek I've replaced all the route names. I was worried that this will break the ability to use controller level routes, but then I thought to add substitutions for area, controller and action when the custom route name is used:
image

This worked when I tested in Razor with `@Url.RouteUrl("AdminMenuList", new { area = "OrchardCore.AdminMenu" })` which correctly returns `/Admin/AdminMenu/List`. (By the way I have never used this method before, but it looks like you have to pass the area too. I think this is normal.) We can have another review when you have the time.

**Comment by MikeAlhayek** [comment]
Well yeah you can't use route name if you are using the attribute on the controller.  In order for the name to work it would have to be placed over each action. Someone may chose to use the attribute in the controller level, but will not have the ability to specify a route name.


I think you have to update the CMS module template with this attribute as well.  Lastly, please update the documents with the change. At least the 1.9 release notes and maybe where ever we have documented the AdminAttribute.  I'll try to do another round of reviews in the morning

**Comment by sarahelsaig** [comment]
Followup to my previous comment, I have tried the same `Url.RouteUrl` in the `main` branch and the results are the same. So you already had to add the "area" route value before my contrib.

> Well yeah you can't use route name if you are using the attribute on the controller. In order for the name to work it would have to be placed over each action. Someone may chose to use the attribute in the controller level, but will not have the ability to specify a route name. Unless you are able to tell that the attribute is at the controller level and append the action name to the provided route name "if any". Route names must be unique

Don't worry it actually works. The `AdminAreaControllerRouteMapper` applies one route per action. That's why it was already doing replace on `"{action}"` [in the `main` branch](https://github.com/OrchardCMS/OrchardCore/blob/adb3b9b986a408a2911264c57b04f1a0efc00d01/src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs#L31). So by using route names such as `"AdminMenu{action}"` and letting the route mapper do the same kind of substitution on the name as well, routes named `AdminMenuList`, `AdminMenuCreate`, etc are created as expected.

> I think you have to update the CMS module template with this attribute as well. 

Could you elaborate? I've checked `OrchardCore.Templates.Cms.Module` but [the only route mapping in it](https://github.com/OrchardCMS/OrchardCore/blob/cf4662ab48b54158fbef04b8719622c40cfba2f0/src/Templates/OrchardCore.ProjectTemplates/content/OrchardCore.Templates.Cms.Module/Startup.cs#L44-L48) is not related to admin so I couldn't see anything that should be updated there.

> Lastly, please update the documents with the change. At least the 1.9 release notes and maybe where ever we have documented the AdminAttribute.

I've added it to the Change Logs. I didn't find any other documentation mentioning the AdminAttribute.

> I'll try to do another round of reviews in the morning

Thank you!

File: src/OrchardCore.Modules/OrchardCore.ContentTypes/Controllers/AdminController.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.ContentTypes/Controllers/AdminController.cs:451]: new. But good

File: src/docs/releases/1.9.0.md
Comment (MikeAlhayek) [inline src/docs/releases/1.9.0.md:161]: Let's not use `.` in the example.

**Review by MikeAlhayek — APPROVED** [review]
@sarahelsaig Lots of work done here. LGTM. Great Job!

I think you should drop the ` NameFromControllerAndActionDotSeparated` constant so you don't encourage people using it. Using one `NameFromControllerAndAction` will keep routes name constancy in case someone wants to use it,

**Comment by sarahelsaig** [comment]
@MikeAlhayek Good idea, I've removed it and updated the example.

File: src/OrchardCore/OrchardCore.Admin.Abstractions/AdminAttribute.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Admin.Abstractions/AdminAttribute.cs:36]:
```suggestion
        /// Gets or sets the name used when mapping the admin route. You can use <c>{area}</c>, <c>{controller}</c>,
        /// <c>{action}</c> in the string, which will be substituted. For performance reasons, these values are exact
        /// and shouldn't contain spaces. If empty when <see cref="Template"/> is used, the fallback is <see
        /// cref="ControllerActionDescriptor.DisplayName"/>.
```

File: src/docs/releases/1.9.0.md
Comment (Piedone) [inline src/docs/releases/1.9.0.md:158]: Just to clarify that it doesn't simply prepend `/Admin`.

```suggestion
The `[Admin]` attribute now has optional parameters for a custom route template and route name. It works just like the `[Route(template, name)]` attribute, except it prepends the configured admin prefix. You can apply it to the controller or the action; if both are specified then the action's template takes precedence. The route name can contain `{area}`, `{controller}`, and `{action}`, which are substituted during mapping so the names can be unique for each action. This means you don't have to define these admin routes in your module's `Startup` class anymore, but that option is still available and supported. Take a look at this example:
```

File: src/docs/releases/1.9.0.md
Comment (Piedone) [inline src/docs/releases/1.9.0.md:173]:
```suggestion
In this example, (if the admin prefix remains the default "Admin") you can reach the Index action at `~/Admin/Person` (or by the route name `Person`), because its own action-level attribute took precedence. You can reach Create at `~/Admin/Person/Create` (route name `PersonCreate`) and Edit for the person whose identifier string is "john-doe" at `~/Admin/Person/john-doe` (route name `PersonEdit`).
```

File: src/docs/releases/1.9.0.md
Comment (Piedone) [inline src/docs/releases/1.9.0.md:156]: Please add this section to https://docs.orchardcore.net/en/latest/docs/reference/modules/Admin/ too.

File: src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:1]: The `Startup`s in `OrchardCore.Admin`, `OrchardCore.AdminDashboard`, `OrchardCore.Demo`, and `OrchardCore.OpenId` remain. In Demo one can remain to show that way of doing it too.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs
Comment (sarahelsaig) [reply src/OrchardCore.Modules/OrchardCore.Admin/AdminAreaControllerRouteMapper.cs:1]: In admin and admin dashboard the route is just the admin prefix, so those can't be replaced this way because the route would be an empty string.

**Comment by sarahelsaig** [comment]
I've addressed your previous feedback so whenever you have time for it.

**Comment by Piedone** [comment]
Please re-request review in such cases so reviewers can know:

image

File: src/OrchardCore.Modules/OrchardCore.Demo/Startup.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Demo/Startup.cs:67]: @Piedone this is a demo module not a training module. This would be a great add to the training module and the docs for reference. Not sure we should have it as part of the framework. But, if you still think we should keep it, I am good with it.

**Review by MikeAlhayek — APPROVED** [review]

File: src/OrchardCore.Modules/OrchardCore.Demo/Startup.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Demo/Startup.cs:67]: Why would a comment hurt here though? It demonstrates two approaches; it's useful to have an explanation of why there are two.

**Pull request merged**
